src/net_processing.cpp

```diff
@@ -1264,6 +1264,9 @@ bool static ProcessHeadersMessage(CNode *pfrom, CConnman *connman, const std::ve
             if (nDoS > 0) {
                 LOCK(cs_main);
                 Misbehaving(pfrom->GetId(), nDoS);
+            } else if (!pfrom->fInbound && !pfrom->m_manual_connection && punish_invalid) {
+                // Don't keep outbound peers that are on an invalid chain
```

Comment:
It's weird to have a punish_invalid param that only punishes based on its own criteria. That means that both sites have to consider all constraints.

Instead, from the ::HEADERS handler, how about something like:
```c++
bool punish_invalid = !pfrom->fInbound && !pfrom->m_manual_connection;
ProcessHeadersMessage(..., should_punish);
```

Reply:
Sounds good.